File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.java
import java.util.*;

class Solution {
    public int solution(int[] topping) {
        int answer = 0;
        Map<Integer, Integer> first = new HashMap<>();
        Set<Integer> second = new HashSet<>();
        
        for(int i = 0; i < topping.length; i++){
            first.put(topping[i], first.getOrDefault(topping[i], 0) + 1);
        }
        
        for(int i = 0; i < topping.length; i++){
            second.add(topping[i]);
            first.put(topping[i], first.get(topping[i]) - 1);
            if(first.get(topping[i]) == 0){
                first.remove(topping[i]);
            }
            if(second.size() == first.size()){
                answer++;
            }
            
        }
        
        return answer;
    }
}

File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.swift
import Foundation

func solution(_ topping:[Int]) -> Int {
    var answer = 0
    var first = Dictionary<Int, Int>()
    var second = Set<Int>()
    
    for i in topping {
        if let count = first[i] {
            first[i] = count + 1
        } else {
            first[i] = 1
        }
    }
    
    for i in topping {
        second.insert(i)
        if let count = first[i] {
            first[i] = count - 1
        }
        if let count = first[i], count == 0 {
            first[i] = nil
        }
        if first.count == second.count {
            answer += 1
        }
    }
    
    
    return answer
}